File: python/samples/queue_processing/mongodb_queue_processor.py
"""
MongoDB Queue Processor with Bulk Insert Operations

Demonstrates efficient queue processing with MongoDB bulk inserts:
- Batch consumption from queue
- Bulk insert operations
- Automatic retry logic
- Performance monitoring
"""

from pymongo import MongoClient, InsertOne
from pymongo.errors import BulkWriteError
from queue import Queue
from typing import List, Dict, Any
import time
import threading


class MongoDBQueueProcessor:
    """
    Process queued items with MongoDB bulk insert operations
    
    Features:
    - Configurable batch sizes
    - Automatic retry on failures
    - Connection pooling
    - Performance metrics
    """
    
    def __init__(
        self,
        connection_string: str = "mongodb://localhost:27017",
        database_name: str = "market_data",
        collection_name: str = "orders"
    ):
        """
        Initialize MongoDB queue processor
        
        Args:
            connection_string: MongoDB connection string
            database_name: Target database name
            collection_name: Target collection name
        """
        self.client = MongoClient(connection_string)
        self.db = self.client[database_name]
        self.collection = self.db[collection_name]
        self.queue = Queue()
        self.processed_count = 0
        self.error_count = 0
        self.total_time = 0
        
    def enqueue(self, item: Dict[str, Any]) -> None:
        """Add item to processing queue"""
        self.queue.put(item)
        
    def enqueue_batch(self, items: List[Dict[str, Any]]) -> None:
        """Add multiple items to queue"""
        for item in items:
            self.queue.put(item)
            
    def bulk_insert(self, documents: List[Dict[str, Any]]) -> int:
        """
        Perform bulk insert operation
        
        Args:
            documents: List of documents to insert
            
        Returns:
            Number of documents inserted
        """
        if not documents:
            return 0
            
        try:
            start_time = time.time()
            
            # Use bulk_write for better performance
            operations = [InsertOne(doc) for doc in documents]
            result = self.collection.bulk_write(operations, ordered=False)
            
            elapsed = time.time() - start_time
            self.total_time += elapsed
            self.processed_count += result.inserted_count
            
            return result.inserted_count
            
        except BulkWriteError as e:
            # Some documents may have been inserted
            self.processed_count += e.details.get('nInserted', 0)
            self.error_count += len(e.details.get('writeErrors', []))
            print(f"Bulk write error: {len(e.details.get('writeErrors', []))} errors")
            return e.details.get('nInserted', 0)
            
    def single_insert(self, document: Dict[str, Any]) -> bool:
        """
        Insert single document (for small queues)
        
        Args:
            document: Document to insert
            
        Returns:
            True if successful
        """
        try:
            start_time = time.time()
            self.collection.insert_one(document)
            self.total_time += time.time() - start_time
            self.processed_count += 1
            return True
        except Exception as e:
            self.error_count += 1
            print(f"Insert error: {e}")
            return False
            
    def process_queue(
        self,
        batch_size: int = 1000,
        workers: int = 1,
        adaptive: bool = True
    ) -> None:
        """
        Process queue with optional adaptive batching
        
        Args:
            batch_size: Maximum batch size for bulk operations
            workers: Number of worker threads
            adaptive: Enable adaptive batching based on queue size
        """
        def worker():
            while True:
                batch = []
                
                # Determine batch size based on queue depth
                if adaptive:
                    queue_size = self.queue.qsize()
                    if queue_size < 100:
                        # Small queue: process individually
                        current_batch_size = 1
                    elif queue_size < 1000:
                        # Medium queue: small batches
                        current_batch_size = 100
                    else:
                        # Large queue: full batches
                        current_batch_size = batch_size
                else:
                    current_batch_size = batch_size
                    
                # Collect items for batch
                for _ in range(current_batch_size):
                    try:
                        item = self.queue.get(timeout=1)
                        batch.append(item)
                        self.queue.task_done()
                    except:
                        break
                        
                if not batch:
                    continue
                    
                # Process batch
                if len(batch) == 1:
                    self.single_insert(batch[0])
                else:
                    self.bulk_insert(batch)
                    
        # Start worker threads
        threads = []
        for _ in range(workers):
            t = threading.Thread(target=worker, daemon=True)
            t.start()
            threads.append(t)
            
        # Wait for queue to be empty
        self.queue.join()
        
    def get_stats(self) -> Dict[str, Any]:
        """
        Get processing statistics
        
        Returns:
            Dictionary with performance metrics
        """
        throughput = (
            self.processed_count / self.total_time
            if self.total_time > 0
            else 0
        )
        
        return {
            "processed_count": self.processed_count,
            "error_count": self.error_count,
            "total_time_seconds": round(self.total_time, 2),
            "throughput_per_second": round(throughput, 2),
            "queue_size": self.queue.qsize()
        }
        
    def close(self) -> None:
        """Close database connection"""
        self.client.close()


# Example usage and demonstration
if __name__ == "__main__":
    import random
    
    print("=== MongoDB Queue Processor Demo ===\n")
    
    # Initialize processor
    processor = MongoDBQueueProcessor(
        connection_string="mongodb://localhost:27017",
        database_name="market_data",
        collection_name="orders"
    )
    
    # Generate sample market orders
    print("Generating sample market orders...")
    symbols = ["AAPL", "GOOGL", "MSFT", "AMZN", "TSLA"]
    
    for i in range(5000):
        order = {
            "order_id": f"ORD{i:06d}",
            "symbol": random.choice(symbols),
            "quantity": random.randint(1, 1000),
            "price": round(random.uniform(100, 500), 2),
            "side": random.choice(["BUY", "SELL"]),
            "timestamp": time.time(),
            "order_type": "MARKET"
        }
        processor.enqueue(order)
        
    print(f"Enqueued {processor.queue.qsize()} orders")
    
    # Process with adaptive batching
    print("\nProcessing queue with adaptive batching...")
    start = time.time()
    processor.process_queue(batch_size=1000, workers=4, adaptive=True)
    
    # Display stats
    print(f"\nProcessing completed in {time.time() - start:.2f} seconds")
    print(f"Statistics: {processor.get_stats()}")
    
    processor.close()
    print("\nDemo completed!")

File: javascript/nodejs/queue_processing/mongodb-queue-processor.js
/**
 * MongoDB Queue Processor with Bulk Insert Operations
 * 
 * Demonstrates efficient queue processing with MongoDB bulk inserts:
 * - Batch consumption from queue
 * - Bulk insert operations
 * - Automatic retry logic
 * - Performance monitoring
 */

const { MongoClient } = require('mongodb');

class MongoDBQueueProcessor {
  /**
   * Initialize MongoDB queue processor
   * @param {Object} options - Configuration options
   * @param {string} options.connectionString - MongoDB connection string
   * @param {string} options.databaseName - Target database name
   * @param {string} options.collectionName - Target collection name
   */
  constructor({
    connectionString = 'mongodb://localhost:27017',
    databaseName = 'market_data',
    collectionName = 'orders'
  } = {}) {
    this.connectionString = connectionString;
    this.databaseName = databaseName;
    this.collectionName = collectionName;
    this.queue = [];
    this.processedCount = 0;
    this.errorCount = 0;
    this.totalTime = 0;
    this.client = null;
    this.db = null;
    this.collection = null;
  }

  /**
   * Connect to MongoDB
   */
  async connect() {
    if (!this.client) {
      this.client = new MongoClient(this.connectionString);
      await this.client.connect();
      this.db = this.client.db(this.databaseName);
      this.collection = this.db.collection(this.collectionName);
    }
  }

  /**
   * Add item to processing queue
   * @param {Object} item - Item to enqueue
   */
  enqueue(item) {
    this.queue.push(item);
  }

  /**
   * Add multiple items to queue
   * @param {Array} items - Items to enqueue
   */
  enqueueBatch(items) {
    this.queue.push(...items);
  }

  /**
   * Perform bulk insert operation
   * @param {Array} documents - Documents to insert
   * @returns {number} Number of documents inserted
   */
  async bulkInsert(documents) {
    if (!documents || documents.length === 0) {
      return 0;
    }

    await this.connect();

    try {
      const startTime = Date.now();

      // Use bulkWrite for better performance
      const operations = documents.map(doc => ({
        insertOne: { document: doc }
      }));

      const result = await this.collection.bulkWrite(operations, {
        ordered: false
      });

      const elapsed = (Date.now() - startTime) / 1000;
      this.totalTime += elapsed;
      this.processedCount += result.insertedCount;

      return result.insertedCount;
    } catch (error) {
      if (error.writeErrors) {
        // Some documents may have been inserted
        this.processedCount += error.result?.nInserted || 0;
        this.errorCount += error.writeErrors.length;
        console.error(`Bulk write error: ${error.writeErrors.length} errors`);
        return error.result?.nInserted || 0;
      }
      throw error;
    }
  }

  /**
   * Insert single document (for small queues)
   * @param {Object} document - Document to insert
   * @returns {boolean} True if successful
   */
  async singleInsert(document) {
    await this.connect();

    try {
      const startTime = Date.now();
      await this.collection.insertOne(document);
      this.totalTime += (Date.now() - startTime) / 1000;
      this.processedCount++;
      return true;
    } catch (error) {
      this.errorCount++;
      console.error('Insert error:', error.message);
      return false;
    }
  }

  /**
   * Process queue with optional adaptive batching
   * @param {Object} options - Processing options
   * @param {number} options.batchSize - Maximum batch size
   * @param {number} options.workers - Number of worker threads (simplified in single-threaded Node.js)
   * @param {boolean} options.adaptive - Enable adaptive batching
   */
  async processQueue({
    batchSize = 1000,
    workers = 1,
    adaptive = true
  } = {}) {
    while (this.queue.length > 0) {
      // Determine batch size based on queue depth
      let currentBatchSize;
      if (adaptive) {
        const queueSize = this.queue.length;
        if (queueSize < 100) {
          currentBatchSize = 1;
        } else if (queueSize < 1000) {
          currentBatchSize = 100;
        } else {
          currentBatchSize = batchSize;
        }
      } else {
        currentBatchSize = batchSize;
      }

      // Extract batch from queue
      const batch = this.queue.splice(0, currentBatchSize);

      if (batch.length === 0) {
        break;
      }

      // Process batch
      if (batch.length === 1) {
        await this.singleInsert(batch[0]);
      } else {
        await this.bulkInsert(batch);
      }
    }
  }

  /**
   * Get processing statistics
   * @returns {Object} Performance metrics
   */
  getStats() {
    const throughput = this.totalTime > 0
      ? this.processedCount / this.totalTime
      : 0;

    return {
      processedCount: this.processedCount,
      errorCount: this.errorCount,
      totalTimeSeconds: parseFloat(this.totalTime.toFixed(2)),
      throughputPerSecond: parseFloat(throughput.toFixed(2)),
      queueSize: this.queue.length
    };
  }

  /**
   * Close database connection
   */
  async close() {
    if (this.client) {
      await this.client.close();
      this.client = null;
    }
  }
}

// Example usage and demonstration
async function main() {
  console.log('=== MongoDB Queue Processor Demo ===\n');

  // Initialize processor
  const processor = new MongoDBQueueProcessor({
    connectionString: 'mongodb://localhost:27017',
    databaseName: 'market_data',
    collectionName: 'orders'
  });

  // Generate sample market orders
  console.log('Generating sample market orders...');
  const symbols = ['AAPL', 'GOOGL', 'MSFT', 'AMZN', 'TSLA'];

  for (let i = 0; i < 5000; i++) {
    const order = {
      order_id: `ORD${String(i).padStart(6, '0')}`,
      symbol: symbols[Math.floor(Math.random() * symbols.length)],
      quantity: Math.floor(Math.random() * 1000) + 1,
      price: parseFloat((Math.random() * 400 + 100).toFixed(2)),
      side: Math.random() > 0.5 ? 'BUY' : 'SELL',
      timestamp: Date.now() / 1000,
      order_type: 'MARKET'
    };
    processor.enqueue(order);
  }

  console.log(`Enqueued ${processor.queue.length} orders`);

  // Process with adaptive batching
  console.log('\nProcessing queue with adaptive batching...');
  const start = Date.now();
  await processor.processQueue({ batchSize: 1000, workers: 4, adaptive: true });

  // Display stats
  const elapsed = (Date.now() - start) / 1000;
  console.log(`\nProcessing completed in ${elapsed.toFixed(2)} seconds`);
  console.log('Statistics:', processor.getStats());

  await processor.close();
  console.log('\nDemo completed!');
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = { MongoDBQueueProcessor };
